platforms/core-execution/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherProbeRegistry.java

```diff
@@ -16,97 +16,172 @@
 
+import javax.annotation.Nullable;
 import java.io.File;
+import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.nio.file.FileStore;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.function.Function;
+import java.util.function.BiConsumer;
 import java.util.stream.Stream;
 
 public class DefaultFileWatcherProbeRegistry implements FileWatcherProbeRegistry {
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultFileWatcherProbeRegistry.class);
 
-    private final Map<File, WatchProbe> watchProbesByHierarchy = new ConcurrentHashMap<>();
-    private final Map<String, WatchProbe> watchProbesByPath = new ConcurrentHashMap<>();
+    private final Map<FileStore, WatchProbeImpl> probesByFS = new ConcurrentHashMap<>();
+    private final Map<Path, WatchProbeImpl> watchProbesByHierarchy = new ConcurrentHashMap<>();
+    private final Set<File> unwatchedHierarchies = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    private ImmutableSet<File> probedHierarchies = ImmutableSet.of();
 
-    private final Function<File, File> probeLocationResolver;
+    @Override
+    public void registerProbe(File watchableHierarchy, @Nullable File probeLocation) {
+        LOGGER.debug("Registering probe for {}", watchableHierarchy);
+
+        FileStore fileStore = getFileStore(watchableHierarchy.toPath());
+        WatchProbeImpl watchProbe;
+        if (fileStore != null) {
+            watchProbe = probesByFS.computeIfAbsent(fileStore, fs -> {
+                File probeFile = getProbeFile(probeLocation, watchableHierarchy);
+                return new WatchProbeImpl(probeFile, isSubpath(probeFile, watchableHierarchy));
+            });
+        } else { // fallback, unlikely to happen
+            File probeFile = getProbeFile(probeLocation, watchableHierarchy);
+            watchProbe = watchProbesByHierarchy
+                .values()
+                .stream()
+                .filter(probe -> probe.probeFile.equals(probeFile))
+                .findFirst()
+                .orElseGet(() -> new WatchProbeImpl(probeFile, isSubpath(probeFile, watchableHierarchy)));
 
-    public DefaultFileWatcherProbeRegistry(Function<File, File> probeLocationResolver) {
-        this.probeLocationResolver = probeLocationResolver;
+            watchProbesByHierarchy.put(watchableHierarchy.toPath(), watchProbe);
+        }
+
+        watchProbe.addWatchableHierarchy(watchableHierarchy);
+    }
+
+    private static File getProbeFile(@Nullable File probeLocation, File watchableHierarchy) {
+        if (probeLocation == null) {
+            probeLocation = new File(watchableHierarchy, ".gradle");
+        }
+        return new File(probeLocation, "file-system.probe");
+    }
```

Comment:
❓ What is the advantage of passing around `null` and then resolving the probe file to this default here, as opposed to resolving it up front, and always passing a non-null `File` around?

Reply:
It caused by the CC code here
https://github.com/gradle/gradle/pull/28611#discussion_r1587195801
It simply is not aware of the probe location.

If that code hasn't existed, then ProbeLocation would always be non-null.